scripts/normalize-by-median.py

```diff
@@ -105,30 +130,67 @@ def normalize_by_median(input_filename, outfp, htable, paired, cutoff,
             discarded += batch_size
 
     if report_fp:
-        print >> report_fp, total, total - discarded, \
-            1. - (discarded / float(total))
+        print(str(total) + " " + str(total - discarded) + " " +
+              str(1. - (discarded / float(total))), file=report_fp)
         report_fp.flush()
 
     return total, discarded
 
 
 def handle_error(error, output_name, input_name, fail_save, htable):
-    print >> sys.stderr, '** ERROR:', error
-    print >> sys.stderr, '** Failed on {name}: '.format(name=input_name)
+    print('** ERROR: ' + str(error), file=sys.stderr)
+    print('** Failed on {name}: '.format(name=input_name), file=sys.stderr)
     if fail_save:
         tablename = os.path.basename(input_name) + '.ct.failed'
-        print >> sys.stderr, \
-            '** ...dumping k-mer counting table to {tn}'.format(tn=tablename)
+        print('** ...dumping k-mer counting table to {tn}'
+              .format(tn=tablename), file=sys.stderr)
         htable.save(tablename)
     try:
         os.remove(output_name)
     except:  # pylint: disable=bare-except
-        print >> sys.stderr, '** ERROR: problem removing corrupt filtered file'
+        print('** ERROR: problem removing corrupt filtered file',
+              file=sys.stderr)
+
+
+@contextmanager
+def CatchIOErrors(ifile, ofile, save_on_fail, ht, force):
+    """
+    Context manager to do boilerplate excepting of IOErrors; also does
+    upkeep on some statistics and diagnostic output.
+    """
+
+    global corrupt_files, total, discarded, total_acc, discarded_acc
+    try:
+        yield
+    except IOError as err:
+        handle_error(err, ofile, ifile, save_on_fail,
+                     ht)
+        if not force:
+            print >> sys.stderr, '** Exiting!'
+
+            sys.exit(1)
+        else:
+            print('*** Skipping error file, moving on...', file=sys.stderr)
+            corrupt_files.append(ifile)
+
+    if total_acc == 0 and discarded_acc == 0:
+        print('SKIPPED empty file ' + ifile, file=sys.stderr)
+    elif total_acc is not None and discarded_acc is not None:
+        # will be none if the context manager caught an error
```

Comment:
can errors be passed other ways? especially if we are making the context manager into an object? this is a bit non-intuitive/complex.